File: android/app/src/main/java/com/slowthemdown/android/debug/SeedData.kt
package com.slowthemdown.android.debug

import android.content.Context
import com.slowthemdown.android.data.db.SpeedEntryDao
import com.slowthemdown.android.data.db.SpeedEntryEntity
import com.slowthemdown.shared.model.CalibrationMethod
import com.slowthemdown.shared.model.TravelDirection
import com.slowthemdown.shared.model.VehicleType
import kotlinx.coroutines.flow.first
import java.util.Calendar
import kotlin.random.Random

object SeedData {

    private const val PREFS_NAME = "seed_data_prefs"
    private const val KEY_SEEDED = "isDemoDataSeeded"

    fun isSeeded(context: Context): Boolean =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .getBoolean(KEY_SEEDED, false)

    private fun setSeeded(context: Context, value: Boolean) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit().putBoolean(KEY_SEEDED, value).apply()
    }
    private val streets = listOf(
        "Oak Street", "Maple Avenue", "Elm Drive", "Pine Road",
        "Cedar Lane", "Birch Way", "Walnut Street", "Cherry Blvd",
        "Willow Court", "Spruce Terrace",
    )

    private val vehicleTypes = VehicleType.entries.toList()
    private val directions = TravelDirection.entries.toList()
    // Speed limits in m/s (25, 25, 25, 30, 30, 35 MPH)
    private val speedLimits = listOf(11.176, 11.176, 11.176, 13.4112, 13.4112, 15.6464)

    fun generate(count: Int = 50): List<SpeedEntryEntity> {
        val calendar = Calendar.getInstance()
        val now = System.currentTimeMillis()

        return (0 until count).map { i ->
            val daysAgo = Random.nextInt(0, 31)
            val hour = Random.nextInt(6, 23)
            val minute = Random.nextInt(0, 60)

            calendar.timeInMillis = now
            calendar.add(Calendar.DAY_OF_YEAR, -daysAgo)
            calendar.set(Calendar.HOUR_OF_DAY, hour)
            calendar.set(Calendar.MINUTE, minute)
            val timestamp = calendar.timeInMillis

            val roll = Random.nextDouble()
            // Base speeds in MPH, then convert to m/s
            val baseSpeedMph = when {
                roll < 0.15 -> Random.nextDouble(15.0, 20.0)
                roll < 0.70 -> Random.nextDouble(22.0, 30.0)
                roll < 0.90 -> Random.nextDouble(30.0, 38.0)
                else -> Random.nextDouble(38.0, 50.0)
            }
            val speedMps = baseSpeedMph * 0.44704

            SpeedEntryEntity(
                timestamp = timestamp,
                speed = Math.round(speedMps * 100.0) / 100.0,
                speedLimit = speedLimits.random(),
                streetName = streets[i % streets.size],
                notes = if (i % 7 == 0) "School zone" else "",
                vehicleTypeRaw = vehicleTypes.random().rawValue,
                directionRaw = directions.random().rawValue,
                calibrationMethodRaw = CalibrationMethod.MANUAL_DISTANCE.rawValue,
                timeDeltaSeconds = Random.nextDouble(0.2, 1.5),
                pixelDisplacement = Random.nextDouble(100.0, 500.0),
                pixelsPerMeter = 98.425, // ~30 px/ft * 3.28084
                referenceDistanceMeters = 3.048, // 10 ft
                latitude = 37.7749 + Random.nextDouble(-0.01, 0.01),
                longitude = -122.4194 + Random.nextDouble(-0.01, 0.01),
            )
        }
    }

    suspend fun seedIfEmpty(dao: SpeedEntryDao, context: Context) {
        val count = dao.getCount().first()
        if (count == 0) {
            generate().forEach { dao.insert(it) }
            setSeeded(context, true)
        }
    }

    suspend fun clearDemoData(dao: SpeedEntryDao, context: Context) {
        dao.deleteAll()
        setSeeded(context, false)
    }
}

File: ios/SlowThemDown/Debug/SeedData.swift
import Foundation
import SwiftData

enum SeedData {
    private static let seededKey = "isDemoDataSeeded"

    static var isSeeded: Bool {
        UserDefaults.standard.bool(forKey: seededKey)
    }

    static func clearDemoData(context: ModelContext) {
        let descriptor = FetchDescriptor<SpeedEntry>()
        if let entries = try? context.fetch(descriptor) {
            for entry in entries {
                context.delete(entry)
            }
        }
        UserDefaults.standard.set(false, forKey: seededKey)
    }

    #if DEBUG
    static let streets = [
        "Oak Street", "Maple Avenue", "Elm Drive", "Pine Road",
        "Cedar Lane", "Birch Way", "Walnut Street", "Cherry Blvd",
        "Willow Court", "Spruce Terrace",
    ]

    static func generate(count: Int = 50) -> [SpeedEntry] {
        var entries: [SpeedEntry] = []
        let calendar = Calendar.current

        for i in 0..<count {
            // Spread entries over the past 30 days
            let daysAgo = Int.random(in: 0...30)
            let hour = Int.random(in: 6...22)
            let minute = Int.random(in: 0...59)

            var components = calendar.dateComponents([.year, .month, .day], from: .now)
            components.day! -= daysAgo
            components.hour = hour
            components.minute = minute
            let timestamp = calendar.date(from: components) ?? .now

            // Realistic speed distribution: mostly 20-35, some outliers
            let baseSpeed: Double
            let roll = Double.random(in: 0...1)
            if roll < 0.15 {
                baseSpeed = Double.random(in: 15...20) // slow
            } else if roll < 0.70 {
                baseSpeed = Double.random(in: 22...30) // normal
            } else if roll < 0.90 {
                baseSpeed = Double.random(in: 30...38) // fast
            } else {
                baseSpeed = Double.random(in: 38...50) // speeder
            }

            let speedLimit = [25, 25, 25, 30, 30, 35].randomElement()!
            let vehicleType = VehicleType.allCases.randomElement()!
            let direction = TravelDirection.allCases.randomElement()!
            let street = streets[i % streets.count]

            let entry = SpeedEntry(
                speedMPH: round(baseSpeed * 10) / 10,
                speedLimit: speedLimit,
                streetName: street,
                notes: i % 7 == 0 ? "School zone" : "",
                vehicleType: vehicleType,
                direction: direction,
                calibrationMethod: .manualDistance,
                timeDeltaSeconds: Double.random(in: 0.2...1.5),
                pixelDisplacement: Double.random(in: 100...500),
                pixelsPerFoot: 30.0,
                referenceDistanceFeet: 10.0,
                latitude: 37.7749 + Double.random(in: -0.01...0.01),
                longitude: -122.4194 + Double.random(in: -0.01...0.01),
                timestamp: timestamp
            )
            entries.append(entry)
        }
        return entries
    }

    static func seedIfEmpty(context: ModelContext) {
        let descriptor = FetchDescriptor<SpeedEntry>()
        let count = (try? context.fetchCount(descriptor)) ?? 0
        if count == 0 {
            let entries = generate()
            for entry in entries {
                context.insert(entry)
            }
            UserDefaults.standard.set(true, forKey: seededKey)
        }
    }
    #endif
}
